src/PackagePrivate/Util/ReplicaMasterAwareRecordIdsAcquirer.php

```diff
@@ -0,0 +1,204 @@
+	 * their values as strings. If some values are of different type (e.g. integer ids)
+	 * this can cause infinite loops due to mismatch in identifying records selected in
+	 * database with their corresponding needed records. The first element keys will be
+	 * used as the set of columns to select in database and to provide back in the returned array.
+	 *
+	 * @param array $neededRecords array of records to be looked-up or inserted.
+	 *	Each entry in this array should an associative array of column => value pairs.
+	 *	Example:
+	 *	[
+	 *		[ 'columnA' => 'valueA1', 'columnB' => 'valueB1' ],
+	 *		[ 'columnA' => 'valueA2', 'columnB' => 'valueB2' ],
+	 *		...
+	 *	]
+	 *
+	 * @return array the array of input recrods along with their ids
+	 *	Example:
+	 *	[
+	 *		[ 'columnA' => 'valueA1', 'columnB' => 'valueB1', 'idColumn' => '1' ],
+	 *		[ 'columnA' => 'valueA2', 'columnB' => 'valueB2', 'idColumn' => '2' ],
+	 *		...
+	 *	]
+	 */
+	public function acquireIds( array $neededRecords ) {
+		$existingRecords = $this->findExistingRecords( $this->dbReplica, $neededRecords );
+		$neededRecords = $this->filterNonExistingRecords( $neededRecords, $existingRecords );
+
+		while ( !empty( $neededRecords ) ) {
+			$this->insertNonExistingRecordsIntoMaster( $neededRecords );
+			$existingRecords = array_merge(
+				$existingRecords,
+				$this->findExistingRecords( $this->dbMaster, $neededRecords )
+			);
+			$neededRecords = $this->filterNonExistingRecords( $neededRecords, $existingRecords );
+		}
+
+		return $existingRecords;
+	}
+
+	private function findExistingRecords( IDatabase $db, array $neededRecords ): array {
+		$recordsSelectConditions = array_map( function ( $record ) use ( $db ) {
+			return $db->makeList( $record, IDatabase::LIST_AND );
+		}, $neededRecords );
+
+		/*
+		 * Todo, related to Note 1 on self::acquireIds():
+		 * this class can allow for specifying a different set of columns to select
+		 * and return back from self::acquireIds(). This set of columns can be added as
+		 * an optional argument to self::acquireIds() for instance, the current solution
+		 * in here can be a fallback when that isn't given.
+		 */
+		$selectColumns = array_keys( $neededRecords[0] );
+		$selectColumns[] = $this->idColumn;
+
+		$existingRows = $db->select(
+			$this->table,
+			$selectColumns,
+			$db->makeList( $recordsSelectConditions, IDatabase::LIST_OR )
+		);
+
+		$existingRecords = [];
+		foreach ( $existingRows as $row ) {
+			$existingRecord = [];
+			foreach ( $selectColumns as $column ) {
+				$existingRecord[$column] = $row->$column;
+			}
+			$existingRecords[] = $existingRecord;
+		}
+
+		return $existingRecords;
+	}
+
+	private function insertNonExistingRecordsIntoMaster( array $neededRecords ) {
+		$uniqueRecords = [];
+		foreach ( $neededRecords as $record ) {
+			$recordHash = $this->calcRecordHash( $record );
+			if ( !isset( $uniqueRecords[$recordHash] ) ) {
```

Comment:
You can just set `$uniqueRecords[$recordHash]` regardless… the records are equivalent anyway.

Reply:
that is true